### PR TITLE
fix(datastore): Add DateTime, Date, Time, Int, Float, Enum field in indexes to GraphQL input correctly for mutations

### PR DESCRIPTION
**Pull request by thisisabhash**
*Issue #, if available:* None 

*Description of changes:* Add DateTime, Date, Time, Int, Float, Enum field in indexes to GraphQL input correctly for mutations when it is a part of custom primary key. For e.g. In the schema below, `id`, `dob`, `date`, `time`, `phoneNumber`, `priority`, `height` are part of Primary Key.
```
type Customer5
@model {
  id: ID! @primaryKey(sortKeyFields: ["dob","date","time","phoneNumber","priority","height"])
  dob: AWSDateTime!
  date: AWSDate!
  time: AWSTime!
  phoneNumber: Int!
  priority: Priority!
  height: Float!
  firstName: String
  lastName: String
}
```

*Check points: (check or cross out if not relevant)*

- [x] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
~~- [ ] All integration tests pass~~
~~- [ ] Documentation update for the change if required~~
- [x] PR title conforms to conventional commit style
~~- [ ] If breaking change, documentation/changelog update with migration instructions~~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift [hunk 1]
```diff
@@ -220,4 +221,65 @@ class GraphQLRequestSyncCustomPrimaryKeyTests: XCTestCase {
...
+                    dob
+                    firstName
+                    lastName
+                    updatedAt
+                    __typename
+                    _version
+                    _deleted
+                    _lastChangedAt
+                  }
+                }
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift:249]: For conflict resolution enabled API, we care mostly about create/update/delete mutations for sync to cloud and subscriptions for initial sync process. Looks like we have the delete mutation with version tested, can we also add subscription like onCreate test as a minimum? A stretch goal would be to add test for non-conflict resolution enabled API

Comment (lawmicha) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift:249]: + syncQuery test as well!

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift [hunk 2]
```diff
@@ -220,4 +221,153 @@ class GraphQLRequestSyncCustomPrimaryKeyTests: XCTestCase {
...
+            XCTAssertEqual(input["_version"] as? Int, expectedInput["_version"] as? Int)
+        } catch {
+            XCTFail(error.localizedDescription)
+        }
+    }
+
+    func testOnCreateSubscriptionGraphQLRequestWithDateInPK() throws {
+        do {
+            let datetime = try Temporal.DateTime(iso8601String: "2020-01-20T08:00")
+            let customer = CustomerWithDateInPK(dob: datetime, firstName: "John", lastName: "Doe")
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift:289]: can we pass in model name using the schema instead of the object? The method we're testing `subscription(to:subscriptionType)` does not require the a customer object to be passed in so the set up code should be set up the same way

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelIdDecorator.swift [hunk 3]
```diff
@@ -18,7 +18,15 @@ public struct ModelIdDecorator: ModelBasedGraphQLDocumentDecorator {
...
             for key in customPrimaryKeys {
-                fields[key] = model[key] as? String
+                if let date = model[key] as? Temporal.Date {
+                    fields[key] = date.iso8601String
+                } else if let dateTime = model[key] as? Temporal.DateTime {
+                    fields[key] = dateTime.iso8601String
+                } else if let time = model[key] as? Temporal.Time {
+                    fields[key] = time.iso8601String
+                } else {
+                    fields[key] = model[key] as? String
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelIdDecorator.swift:28]: the previous code assumed that the primary key can only be a string. 
1. are all values of the primary key just the ones you've listed here? (string, datetime, date, time). are there anything else? It doesn't make sense to fix it for one field type but not for others. 
2. based on 1. if it expands to more than (string, datetime, date, time), then is there already some code that allows you to extract the graphQL value out of the field? ie. can we reduce the code down to `fields[keys] = model[key].toGraphQLValue()`

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift (hunk 2 above)
Comment (lawmicha) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift:289]: looks like ur doing this below, use `CustomerWithDateInPK.modelName` instead

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift [hunk 4]
```diff
@@ -14,6 +14,7 @@ class GraphQLRequestSyncCustomPrimaryKeyTests: XCTestCase {
 
     override func setUp() {
         ModelRegistry.register(modelType: CustomerOrder.self)
+        ModelRegistry.register(modelType: CustomerWithDateInPK.self)
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift:17]: nit: would prefer this to be a in a different file like `GraphQLRequestSyncCustomPrimaryKeyWithDateTests`

Comment (lawmicha) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyTests.swift:17]: CustomerOrder was written at a very different time, whereas CustomerWithDateInPK was written with V2 transform. Even though this is a unit test, there's an aspect of this where we will most likely provision the backend to continue verifying integration tests. also if we want to add another CPK use case, most likely it will be in another test class so it scales

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelIdDecorator.swift [hunk 5]
```diff
@@ -61,3 +63,49 @@ public struct ModelIdDecorator: ModelBasedGraphQLDocumentDecorator {
         return document.copy(inputs: inputs)
     }
 }
+
+extension Model {
+
+    func graphQLInputForPrimaryKey(modelFieldName: ModelFieldName) -> String? {
+
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelIdDecorator.swift:70]: let's move this over to `Model+GraphQL` since the other similar method also lives there and then we're not duplicating `getFieldValue` in two places

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelIdDecorator.swift [hunk 6]
```diff
@@ -61,3 +63,49 @@ public struct ModelIdDecorator: ModelBasedGraphQLDocumentDecorator {
         return document.copy(inputs: inputs)
     }
 }
+
+extension Model {
+
+    func graphQLInputForPrimaryKey(modelFieldName: ModelFieldName) -> String? {
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/ModelIdDecorator.swift:69]: comment like
```
/// Retrieve the custom primary key's value used for the GraphQL input.
    /// Only a subset of model types are applicable as custom indexes such as
    /// `date`, `dateTime`, `time`, `enum`, `string`, `double`, and `int`.
```

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyWithMultipleFieldsTests.swift [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    func testDeleteMutationGraphQLRequestWithDateInPK() throws {
+        let customer = CustomerWithMultipleFieldsinPK(dob: Temporal.DateTime.now(),
+                                 date: Temporal.Date.now(),
+                                 time: Temporal.Time.now(),
+                                 phoneNumber: 1_234_567,
+                                 priority: Priority.high,
+                                 height: 6.1,
+                                 firstName: "John",
+                                 lastName: "Doe")
```
Comment (lawmicha) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Model/GraphQLRequest/GraphQLRequestSyncCustomPrimaryKeyWithMultipleFieldsTests.swift:32]: indentation

**Review by lawmicha — APPROVED** [review]
LGTM, great coverage on the possible sort key fields. minor comment to move the method over to existing extension method
